File: patterns/graceful-retry-actions/rules-js/graceful-retry-actions-backoff.js
const MAX_RETRIES_BACKOFF = 5;
const MAX_INTERVAL = 60;

rules.JSRule({
  name: "Graceful Retry LightSwitch with Backoff",
  triggers: [triggers.ItemStateChangeTrigger("SomeTrigger", undefined, "ON")],
  execute: (event) => {
    let retryCount = 0;
    let retryInterval = 5;
    let retryTimer = null;

    const attempt = () => {
      try {
        items.getItem("LightSwitch").sendCommand("ON");
        console.log("Command erfolgreich gesendet!");
        if (retryTimer !== null) retryTimer.cancel();
      } catch (e) {
        retryCount++;
        console.warn("Fehlversuch #" + retryCount);
        if (retryCount < MAX_RETRIES_BACKOFF) {
          retryInterval = Math.min(retryInterval * 2, MAX_INTERVAL);
          console.log("Naechster Versuch in " + retryInterval + " Sekunden");
          retryTimer.reschedule(time.ZonedDateTime.now().plusSeconds(retryInterval));
        } else {
          console.error("Maximale Anzahl an Versuchen erreicht!");
          try {
            items.getItem("NotificationItem").postUpdate("LightSwitch konnte nicht eingeschaltet werden");
          } catch (notifyItemEx) {
            console.warn("Konnte NotificationItem nicht aktualisieren: " + notifyItemEx.message);
          }
          try {
            if (actions.NotificationAction) {
              actions.NotificationAction.sendNotification("admin@example.com", "LightSwitch konnte nicht eingeschaltet werden");
            }
          } catch (notifyEx) {
            console.warn("Cloud-Benachrichtigung nicht verfuegbar: " + notifyEx.message);
          }
        }
      }
    };

    retryTimer = actions.ScriptExecution.createTimer(time.ZonedDateTime.now(), attempt);
  }
});

File: patterns/graceful-retry-actions/rules-python/graceful_retry_actions_backoff.py
import threading

from openhab import rule, Registry, logger
from openhab.actions import NotificationAction
from openhab.triggers import ItemStateChangeTrigger

MAX_RETRIES = 5
INITIAL_INTERVAL = 5
MAX_INTERVAL = 60


def send_command_with_backoff(item_name, command, retries=0, interval=INITIAL_INTERVAL):
    try:
        Registry.getItem(item_name).sendCommand(command)
        logger.info("Command '{}' an {} erfolgreich gesendet!".format(command, item_name))
    except Exception as e:
        retries += 1
        logger.warn("Fehlversuch #{} fuer {}: {}".format(retries, item_name, e))
        if retries < MAX_RETRIES:
            next_interval = min(interval * 2, MAX_INTERVAL)
            logger.info("Naechster Versuch in {} Sekunden".format(next_interval))
            threading.Timer(
                next_interval, send_command_with_backoff, args=(item_name, command, retries, next_interval)
            ).start()
        else:
            logger.error("Maximale Anzahl an Versuchen fuer {} erreicht!".format(item_name))
            NotificationAction.sendNotification("admin@example.com", item_name + " konnte nicht eingeschaltet werden")


@rule(triggers=[ItemStateChangeTrigger("SomeTrigger", state="ON")])
class GracefulRetryActionWithBackoff:
    def execute(self, module, input):
        send_command_with_backoff("LightSwitch", "ON")
